File: ruby/collatz-conjecture/collatz_conjecture.rb
module CollatzConjecture
  def self.steps(number, response = 0)
    raise ArgumentError unless number.positive?
    return response if number == 1

    response += 1
    return steps(number / 2, response) if number.even?
    return steps(number * 3 + 1, response) if number.odd?
  end
end

File: javascript/collatz-conjecture/collatz-conjecture.js
//
// This is only a SKELETON file for the 'Collatz Conjecture' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const steps = (number, stepsTaken = 0) => {
  if(number <= 0) {
    throw(new Error("Only positive numbers are allowed"))
  }

  switch(true) {
    case number === 1:
      return stepsTaken;
    case (number % 2 == 0):
      return steps(number / 2, stepsTaken + 1)
    case (number % 2 != 0):
      return steps(number * 3 + 1, stepsTaken + 1)
  }
};
